src/coreclr/vm/interoputil.cpp

```diff
@@ -1400,23 +1400,25 @@ VOID EnsureComStarted(BOOL fCoInitCurrentThread)
         GC_TRIGGERS;
         MODE_ANY;
         PRECONDITION(GetThreadNULLOk() || !fCoInitCurrentThread);
-        PRECONDITION(g_fEEStarted);
+        PRECONDITION(g_fEEStarted || IsFinalizerThread());
     }
     CONTRACTL_END;
 
     if (g_fComStarted == FALSE)
     {
-        FinalizerThread::GetFinalizerThread()->SetRequiresCoInitialize();
-
         // Attempt to set the thread's apartment model (to MTA by default). May not
         // succeed (if someone beat us to the punch). That doesn't matter (since
         // CLR objects are now apartment agile), we only care that a CoInitializeEx
         // has been performed on this thread by us.
         if (fCoInitCurrentThread)
             GetThread()->SetApartment(Thread::AS_InMTA);
 
-        // set the finalizer event
-        FinalizerThread::EnableFinalization();
+        if (!IsFinalizerThread())
+        {
+            FinalizerThread::GetFinalizerThread()->SetRequiresCoInitialize();
+            // set the finalizer event
+            FinalizerThread::EnableFinalization();
+        }
```

Comment:
What is this trying to do? Is it still needed now that we wait for the finalizer thread to CoInitialize COM during startup?

Reply:
I think all the stuff that tries to CoInitialize the current and Finalizer thread lazily can be removed if this change goes in. 
I think quite a bit of cleanup might be possible even without this change. 

I did not want to go too far beyond what is needed for the fix - jut to minimize the diff, in case we backport this.

Reply:
I agree that there is a lot of cruft that can be cleaned up that is better left for a different PR.

I am not able to reason about whether this `if (!IsFinalizerThread()) { ... }` block is just a useless code or whether it can be actively harmful in some corner cases.

For this PR, can we delete the `if (!IsFinalizerThread())` block, `fCoInitCurrentThread` argument for `EnsureComStarted` and the one call to `EnsureComStarted(FALSE)`?

Reply:
I have added `g_pFinalizerThread->SetApartmentOfUnstartedThread(Thread::AS_InMTA)` to match manual call into CoInitialize, but that results in "lazy coinitialize" pattern forcing finalizer to set itself to MTA in response to setting finalizer thread to MTA.  

I think I can remove both this change and `g_pFinalizerThread->SetApartmentOfUnstartedThread(Thread::AS_InMTA)` to raise fewer questions. 
It is ok to use native API to set apartment state before Thread is attached. The rest is unnecessary and perhaps confusing.

Reply:
I am thinking about performance overhead of this change. We are going to do extra thread switch during startup critical path. It is unavoidable.

However, `EnsureComStarted` that is going to be executed a bit later is going to spin the finalizer thread again by setting `FinalizerThread::EnableFinalization`. 

Can we reduce the startup overhead by setting `g_fComStarted` on finalizer thread once we call CoInitialize and reduce EnsureComStarted to just:
```
    _ASSERTE(g_fComStarted); // COM is expected to be started on finalizer thread during startup
```
?

Reply:
> I am thinking about performance overhead of this change. We are going to do extra thread switch during startup critical path. It is unavoidable.

I've moved launching of finalizer thread a bit earlier. Initializing thread will do some other work before it needs to check if FLS/COM initialization is finished. Hopefully it will be at least partially done, so we'd not need to wait much, if at all.

> Can we reduce the startup overhead by setting `g_fComStarted` on finalizer thread

Good point. Since we will be initializing anyways, why not get some benefit from it.